Add persistent_notification service to the notify platform

KevinCathcart opened the pull request:
## Breaking change

A new `persistent_notification` service has been added to the notify intergration. This could break any users who has registered a notifier with that name.

## Proposed change
Add `persistent_notification` service to notify integration.

This was [requested in a WTH](https://community.home-assistant.io/t/why-the-heck-is-there-no-notify-persistent-notification-service/220160), where some good points are made, especially with respect to notification groups and the alert integration.

The existing services are still needed, both for compatibility, and because you cannot delete/mark as read via the `notify` base integration. Therefore, the implementation of this new service is trivial, merely delegating to the existing service.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

No configuration is available.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

KevinCathcart committed:
Add persistent_notify service as pseudo-platform to notify

KevinCathcart commented on homeassistant/components/persistent_notification/notify.py:
```diff
@@ -0,0 +1,32 @@
+        if ATTR_TITLE in kwargs:
+            payload[ATTR_TITLE] = kwargs[ATTR_TITLE]
+
+        await self.hass.services.async_call(DOMAIN, SERVICE_CREATE, payload)
```
I call the service here via the service registry. I'm not sure if importing the underlying function from __init__.py and using that directly would be better. Thoughts?

MartinHjelmare commented on homeassistant/components/persistent_notification/notify.py:
```diff
@@ -0,0 +1,32 @@
+class PersistentNotifyPlatform(BaseNotificationService):
+    """Implement the notification service for the persistent notification notify platform."""
+
+    hass: HomeAssistant
```
We can remove this. It's not needed.

MartinHjelmare commented on tests/components/persistent_notification/test_notify.py:
```diff
@@ -0,0 +1,51 @@
+from tests.common import get_test_home_assistant
+
+
+class TestPersistentNotify(unittest.TestCase):
```
All new tests must be standalone pytest test functions.

KevinCathcart replied:
Ok fair enough. I based this on the group notify tests, so I'll refactor it to be a simple test function.

KevinCathcart commented:
I've addressed the review remarks so far, except my own of course.

Still need a determination on if the auto-discovery is desired, and if so, should it be considered a minor breaking change.

MartinHjelmare commented on tests/components/persistent_notification/test_notify.py:
```diff
@@ -0,0 +1,28 @@
+    assert notification["status"] == pn.STATUS_UNREAD
+    assert notification["message"] == "Hello"
+    assert notification["title"] == "Test notification"
+    return
```
We can remove the `return` statement.

MartinHjelmare commented on tests/components/persistent_notification/test_notify.py:
```diff
@@ -0,0 +1,28 @@
+    await hass.services.async_call(notify.DOMAIN, pn.NOTIFY_SERVICE_NAME, message)
+    await hass.async_block_till_done()
+
+    notifications = hass.data[pn.DOMAIN]["notifications"]
```
Get the state of the notification in the state machine `hass.states` instead. Don't access integration details in the test.

dgomes commented on homeassistant/components/persistent_notification/notify.py:
```diff
@@ -0,0 +1,28 @@
+
+    async def async_send_message(self, message: str, **kwargs: Any) -> None:
+        """Create persistent notification from message."""
+        assert self.hass
```
it would be odd for this to fail

KevinCathcart replied:
It would be odd, but mypy wants that or something like it, because the base class declares hass as `Optional[HomeAssistant]`, since it gets assigned outside of the constructor, so it is indeed unset for certain early method calls.

dgomes reviewed (approved):
LGTM, using the service registry seams a good idea

KevinCathcart commented:
Anybody want to opine on the autodiscovery that effectively just makes this always available as `notify.persistent_notification`? It feels odd just stealing that name from the user, although admittedly if anybody had a notify service with that name, it is almost certainly some custom thing trying to implement exactly this.

tdejneka commented:
Just a thought but what about simply `notify.persistent`?
It's part of the `notify` domain so the "notification" in its name seems a bit redundant.

KevinCathcart commented:
I seriously thought about that, but wasn’t sure if it was better to keep
the name matching the integration, or to go with the grammatically nicer
one. After all, this is not the only type of notification that can be
considered “persistent”. On the other hand matching the integration name
version makes it soundly horribly redundant.

If you look closely at the code, I actually added an extra constant to make
this trivial (one line) to change if we wanted to go with
notify.persistent. I meant to mention that question earlier, but it must
have slipped my mind.
I’m happy to go either way with this.

tdejneka commented:
> this is not the only type of notification that can be considered “persistent”

That's true. I guess the counter-argument is that the name shouldn't stray too far from the original one. Otherwise users might be misled to believe it is something completely new and unrelated to persistent_notification.

MartinHjelmare commented on homeassistant/components/persistent_notification/__init__.py:
```diff
@@ -28,6 +30,8 @@
 
 EVENT_PERSISTENT_NOTIFICATIONS_UPDATED = "persistent_notifications_updated"
 
+NOTIFY_SERVICE_NAME = DOMAIN
```
We need to ensure that no other platform registers this notifier name before this platform. Maybe by adding a check in the notify integration for this name. That's not ideal from an architecture view but I'm not sure there's a better way at the moment.

KevinCathcart replied:
Yeah. I’m not really a fan of adding such a check. I’d sooner just have the notify integration load this platform directly in its own setup method before it processes the platforms in the yaml. While weird, that feels like less of an architectural violation to me. It is equivalent to the notify platform just adding a hard coded record to the front of its yaml config before processing it, and like all integrations, notify is allowed to know that persistent_notification exists.

Alternatively, instead of hard coding, we could force or allow the user to configure this, and thus specify a name of their choosing, and make it their responsibility to avoid conflicts. If we go that way, the question is how.

We could just let the user specify this by adding it under the notify integration yaml key, in the typical fashion. While it violates the intent of [ADR0007](https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md), it does not technically violate the terms, since both are existing integrations, and no change to the yaml schema is requires, as the existing base schema from notify is all that would be needed.

We could alternatively follow the spirit of ADR0007, and let this be configured as part of the schema for the persistent_notification platform. This could allow us to keep a default name, but allow the user to override it. This feels kind of pointless, as there is no real need to be able to change the name, except to avoid issues if something else uses that name first. However, it is architecturally the cleanest solution to the potential for name conflicts.

What we cannot do is configure this via config flow, as it does not integrate with a device or external service, so [ADR0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md) mandates that this be configured by either yaml or storage collections.

MartinHjelmare replied:
Loading this platform first, before other platforms, in the notify integration sounds ok.

KevinCathcart replied:
Ok. Changed to load it first by prepending it to the list of platform to load that got pulled from YAML.

MartinHjelmare reviewed (approved):
Looks good! I'd like a second pair of eyes.

pvizeli commented on homeassistant/components/notify/__init__.py:
```diff
@@ -275,6 +276,13 @@ async def async_setup_platform(
+            {CONF_NAME: persistent_notification.NOTIFY_SERVICE_NAME},
+        )
+    ]
+
```
That is still not possible. We have strict interface concept which avoid to do that. Solution will be to define this as after dependence over meta data. They is also load with default config.

KevinCathcart replied:
I am not entirely sure what you mean. If you are talking about `after_dependencies`, that is pointless as:
1. `persistent_notification` is in the whitelist of imports allowed without specifying as a dependency. Further more, only a single integration lists it as a dependency, despite multiple others using it in the setup method, in a way that requires it to have already been fully loaded. This is because:
2. `persistent_notification` is always loaded as the second loaded integration, immediately after `homeassistant`. This is done without regard to dependencies, and without regard to the config. 

In case your your objection was to something else, I've pushed up an alternative, that goes back to using `discovery.async_load_platform` approach from `persistent_notification`, using configuration on the `persistent_notification` to allow the user to manually avoid any naming conflicts.

KevinCathcart replied:
If however the above did answer your concerns, I'd be happy to revert to that commit again.

balloob commented on homeassistant/components/persistent_notification/__init__.py:
```diff
@@ -97,6 +117,12 @@ def async_dismiss(hass: HomeAssistant, notification_id: str) -> None:
 async def async_setup(hass: HomeAssistant, config: dict) -> bool:
     """Set up the persistent notification component."""
+    # Process configs
+    if DOMAIN in config:
```
I don't think that we should put anything inside `persistent_notification`. I think that instead we should just add this notification service to `notify/__init__.py` and just always set up as `notify.persistent_notification`.

KevinCathcart commented:
Ok. Redesigned to simply be an additional service in the notify integration as per @balloob's suggestion.

The title of this PR is now no longer correct, since I'm no longer adding a platform, but just a service added to the notify platform. (Admittedly it is a bit of a pseudo-platform, in that usage wise it looks like just another notify platform.) 

I still need to add documentation, and probably should mark this as a (minor) breaking change, since it will presumably break anybody who created a notifier called "persistent_notification".

balloob commented on homeassistant/components/notify/__init__.py:
```diff
@@ -215,6 +224,20 @@ async def async_setup(hass, config):
+            title.hass = hass
+            payload[ATTR_TITLE] = title.async_render()
+
+        await hass.services.async_call(pn.DOMAIN, pn.SERVICE_CREATE, payload)
```
```suggestion
        await hass.services.async_call(pn.DOMAIN, pn.SERVICE_CREATE, payload, blocking=True)
```

balloob committed:
Wait till notification created

balloob committed:
Black

ScottG489 commented:
Should `notification_id` have been added to the schema for this?

Otherwise, it doesn't seem possible to clear a notification created from this notify platform. It doesn't look like you can specify a `notification_id` like `persistent_notification.create`. It also doesn't seem like a documentation page for this notify platform was created (e.g. https://www.home-assistant.io/integrations/notify.tts/), but I'm not clear on if that's a requirement.

Sorry for necro'ing this, but thought I'd post here before making an issue.

MartinHjelmare commented:
Better make an issue. Thanks!
